optimizer/resolver.go

```diff
@@ -0,0 +1,568 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package optimizer
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/sessionctx/db"
+)
+
+// ResolveName resolves table name and column name.
+// It generates ResultFields for ResetSetNode and resolve ColumnNameExpr to a ResultField.
+func ResolveName(node ast.Node, info infoschema.InfoSchema, ctx context.Context) error {
+	defaultSchema := db.GetCurrentSchema(ctx)
+	resolver := nameResolver{Info: info, Ctx: ctx, DefaultSchema: model.NewCIStr(defaultSchema)}
+	node.Accept(&resolver)
+	return resolver.Err
+}
+
+// nameResolver is the visitor to resolve table name and column name.
+// In general, a reference can only refer to information that are available for it.
+// So children elements are visited in the order that previous elements make information
+// available for following elements.
+//
+// During visiting, information are collected and stored in resolverContext.
+// When we enter a subquery, a new resolverContext is pushed to the contextStack, so sub query
+// information can overwrite outer query information. When we look up for a column reference,
+// we look up from top to bottom in the contextStack.
+type nameResolver struct {
+	Info          infoschema.InfoSchema
+	Ctx           context.Context
+	DefaultSchema model.CIStr
+	Err           error
+
+	contextStack []*resolverContext
+}
+
+// resolverContext stores information in a single level of select statement
+// that table name and column name can be resolved.
+type resolverContext struct {
+	/* For Select Statement. */
+	// table map to lookup and check table name conflict.
+	tableMap map[string]int
+	// tableSources collected in from clause.
+	tables []*ast.TableSource
+	// result fields collected in select field list.
+	fieldList []*ast.ResultField
+	// result fields collected in group by clause.
+	groupBy []*ast.ResultField
+
+	// The join node stack is used by on condition to find out
+	// available tables to reference. On condition can only
+	// refer to tables involved in current join.
+	joinNodeStack []*ast.Join
+
+	// When visiting TableRefs, tables in this context are not available
+	// because it is being collected.
+	inTableRefs bool
```

Comment:
inXXX works like flag rather than enum, an element can be `inOrderBy` and `inByItemExpression` at the same time.